### PR TITLE
feat(http): Support custom serializer for http requests

### PR DESCRIPTION
**Pull request by youssef3wi**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #48167

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/common/http/src/request.ts [hunk 1]
```diff
@@ -275,8 +276,17 @@ export class HttpRequest<T> {
   /**
    * Transform the free-form body into a serialized format suitable for
    * transmission to the server.
+   * 
+   * @param xhrFactory The XHR factory includes a function for serializing the body.
    */
-  serializeBody(): ArrayBuffer|Blob|FormData|URLSearchParams|string|null {
+  serializeBody(xhrFactory?: XhrFactory): ArrayBuffer|Blob|FormData|URLSearchParams|string|null {
+    // Check whether has a custom serializer or not.
+    if (xhrFactory && xhrFactory.serialize) {
```
Comment (eneajaho) [inline packages/common/http/src/request.ts:284]:
```suggestion
    if (xhrFactory?.serialize) {
```

File: packages/common/http/src/request.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import { XhrFactory } from '@angular/common/public_api';
```
Comment (JeanMeche) [inline packages/common/http/src/request.ts:9]: Never import from `public_api`. 
```suggestion
import { XhrFactory } from '@angular/common';
```

File: packages/common/http/test/request_spec.ts [hunk 3]
```diff
@@ -10,6 +10,8 @@ import {HttpContext} from '@angular/common/http/src/context';
 import {HttpHeaders} from '@angular/common/http/src/headers';
 import {HttpParams} from '@angular/common/http/src/params';
 import {HttpRequest} from '@angular/common/http/src/request';
+import { XhrFactory } from '@angular/common/public_api';
```
Comment (JeanMeche) [inline packages/common/http/test/request_spec.ts:13]: Same. 
```suggestion
import { XhrFactory } from '@angular/common';
```

File: packages/common/http/test/request_spec.ts [hunk 4]
```diff
@@ -10,6 +10,8 @@ import {HttpContext} from '@angular/common/http/src/context';
...
+import { MockXhrFactoryWithSerializer } from './xhr_mock';
```
Comment (JeanMeche) [inline packages/common/http/test/request_spec.ts:14]: You'll need to run `yarn lint` to fix several linting issues, like here the extra space.

**Comment by JeanMeche** [comment]
The Public APIs are tracked by golden files, since you changed the API surface you will need to run ` yarn bazel run //packages/common:common_api.accept` to update those files.

**Comment by youssef3wi** [comment]
Hello @[JeanMeche](https://github.com/JeanMeche), the CLA was signed yesterday using my principal email, even before I opened the PR.

**Comment by JeanMeche** [comment]
You used 2 different accounts for your commits (hence the failed CLA check), could you sort that out and squash the commits ?

**Comment by youssef3wi** [comment]
I have only one account :)
The squash can be made during the merge.
Capture d’écran 2023-07-17 à 13 37 50

**Comment by eneajaho** [comment]
> The PRs are not merged on Github since there are merge into the Google's monorepo.
> 
> @eneajaho it's your gmail adress that shows up in the commits because of your suggestion !

I've already signed the CLA.

**Comment by markchagers** [comment]
Now what? is this pr blocked because of a noreply email address from one of the contributors?

**Comment by youssef3wi** [comment]
> Now what? is this pr blocked because of a noreply email address from one of the contributors?

I did not receive the Google email to contribute to the platform (CLA).

**Pull request closed without merging**